etl/steps/data/meadow/climate/2024-11-19/total_precipitation.py

```diff
@@ -0,0 +1,167 @@
+    da = da.rio.write_crs("epsg:4326")
+
+    return da
+
+
+def _load_shapefile(file_path: str) -> gpd.GeoDataFrame:
+    log.info("load_shapefile.start")
+    shapefile = gpd.read_file(file_path)
+    return shapefile[["geometry", "WB_NAME"]]
+
+
+def run(dest_dir: str) -> None:
+    # Activates the usage of the global context. Using this option can enhance the performance
+    # of initializing objects in single-threaded applications.
+    pyproj.set_use_global_context(True)  # type: ignore
+
+    #
+    # Load inputs.
+    #
+    # Retrieve snapshot.
+    snap = paths.load_snapshot("total_precipitation.zip")
+
+    # Read data from snapshot
+    da = _load_data_array(snap)
+
+    # Read the shapefile to extract country informaiton
+    snap_geo = paths.load_snapshot("world_bank.zip")
+    shapefile_name = "WB_countries_Admin0_10m/WB_countries_Admin0_10m.shp"
+
+    # Check if the shapefile exists in the ZIP archive
+    with zipfile.ZipFile(snap_geo.path, "r"):
+        # Construct the correct path for Geopandas
+        file_path = f"zip://{snap_geo.path}!/{shapefile_name}"
+
+        # Read the shapefile directly from the ZIP archive
+        shapefile = _load_shapefile(file_path)
+
+    #
+    # Process data.
+    #
+
+    # Initialize an empty dictionary to store the country-wise average precipitation.
+    temp_country = {}
+
+    # Add Global mean precipitation
+    weights = np.cos(np.deg2rad(da.latitude))
+    weights.name = "weights"
+    clim_month_weighted = da.weighted(weights)
+    global_mean = clim_month_weighted.mean(["longitude", "latitude"])
+    temp_country["World"] = global_mean
+
+    # Initialize a list to keep track of small countries where precipitation data extraction fails.
+    small_countries = []
+
+    # Iterate over each row in the shapefile data.
+    for i in tqdm(range(shapefile.shape[0])):
+        # Extract the data for the current row.
+        geometry = shapefile.iloc[i]["geometry"]
+        country_name = shapefile.iloc[i]["WB_NAME"]
+
+        try:
+            # Clip to the bounding box for the country's shape to significantly improve performance.
+            xmin, ymin, xmax, ymax = geometry.bounds
+            clip = da.rio.clip_box(minx=xmin, miny=ymin, maxx=xmax, maxy=ymax)
+
+            # Clip data to the country's shape.
+            # NOTE: if memory is an issue, we could use `from_disk=True` arg
+            clip = clip.rio.clip([mapping(geometry)], shapefile.crs)
+
+            # Calculate weights based on latitude to account for area distortion in latitude-longitude grids.
+            weights = np.cos(np.deg2rad(clip.latitude))
+            weights.name = "weights"
+
+            # Apply the weights to the clipped precipitation data.
+            clim_month_weighted = clip.weighted(weights)
+
+            # Calculate the weighted mean precipitation for the country.
+            # country_weighted_mean = clim_month_weighted.mean(dim=["longitude", "latitude"]).values
```

Comment:
thanks a lot for spotting this! this was actually because I was playing around with other ways of calculating precipitation for each country later on and left this here by mistake - it's not a sum as the actual function is also a weighted mean! I'll remove it